### ISSUE TITLE
Index xrefs in type signatures

### ISSUE THREAD

**Issue by zsol** [zsol/py.wtf#8]
We want to be able to link to symbols within and across packages. Instead of a `Type` being a `string`, we'd want the indexer to emit something like

```python
class Type:
  name: str
  xref: XRef | None
  params: None | Sequence[Type] = None

class XRef:
  pkg: Literal["__std__"] | str | None
  fqname: str
```

Use cases covered:

<details><summary><code>Foo</code></summary>
<code>
Type("Foo", XRef("mypkg", "x.y.Foo"))
</code>
</details>
<details><summary><code>dict[str, Foo]</code></summary>
<pre>
Type(
  "dict",
  XRef("__std__", "dict"),
  params=[
    Type("str", XRef("__std__", "str")),
    Type("Foo", XRef("mypkg", "x.y.Foo")),
  ],
)
</pre>
</details>
<details><summary><code>Callable[[], Mode]</code></summary>
<pre>
Type(
  "Callable",
  XRef("__std__", "typing.Callable"),
  params=[
    Type("", xref=None, params=[]),
    Type("Mode", XRef("black", "black.Mode")),
  ],
)
</pre>
</details>
<details><summary><code>Callable[[Foo], Mode]</code></summary>
<pre>
Type(
  "Callable",
  XRef("__std__", "typing.Callable"),
  params=[
    Type("", xref=None, params=[Type("Foo", XRef("mypkg", "x.y.Foo"))]),
    Type("Mode", XRef("black", "black.Mode")),
  ],
)
</pre>
</details>
<details><summary><code>MysteriousFoo</code> (when the indexer can't determine the source)</summary>
<code>
Type("MysteriousFoo", XRef(None, "x.y.Foo"))
</code>
</details>

**Comment by zsol** [comment]
Not sure what to do with `Foo | Bar` syntactic sugar yet. Maybe translate it back to `typing.Union[Foo, Bar]`?

**Comment by abesto** [comment]
Oooh this looks super promising.

* How careful to we need to be about changes in the JSON format, given what you know / expect about the cost of rebuilding the index? (Separately: in the next weeks, vs once this is "live"?)
* Would this also take care of resolving renamed imports to their "canonical" name? Would we want to *display* the canonical name, or the renamed one? Like, think of `import typing as t`.

> Not sure what to do with Foo | Bar syntactic sugar yet. Maybe translate it back to typing.Union[Foo, Bar]?

I see two reasonable approaches:
* Normalize everything similar into one form. Similar is `typing.List[Foo]` vs `list[Foo]`.
* Add an extra enum field "format" or "flavor" or whatever, that tells the renderer how to serialize this back, such that it matches the original in the code.

I guess the decision is: do we *care* about keeping the original format from the code? I'm leaning towards "no".

**Comment by zsol** [comment]
> How careful do we need to be about changes in the JSON format,

For now, the whole website is pushed atomically, so no problem with push-safety. Regenerating the 8 projects we have takes about 10 seconds (in sequence). No biggie.

Somewhere down the line, the size of the website will exceed reasonable limits, at which points we'll push the data separately from the frontend. Changing the schema will be... more problematic from then on. Then again, the solution is just a reload away :)

> Would this also take care of resolving renamed imports to their "canonical" name? Would we want to display the canonical name, or the renamed one? Like, think of import typing as t.

Ideally `XRef.fqname` would only ever contain canonical names (assuming the indexer can figure it out), so simple aliasing by import will be fine, but `import typing; a = typing.Callable` won't be resolved (at least for now - maybe after I nerdsnipe someone from a type checker team).

**Comment by zsol** [comment]
closed by #75
